src/terminal/adapter/PageManager.cpp

```diff
@@ -0,0 +1,254 @@
+void PageManager::Reset()
+{
+    _activePageNumber = 1;
+    _visiblePageNumber = 1;
+    _buffers = {};
+}
+
+Page PageManager::Get(const til::CoordType pageNumber) const
+{
+    const auto requestedPageNumber = std::min(std::max(pageNumber, 1), MAX_PAGES);
+    auto [visibleBuffer, visibleViewport, isMainBuffer] = _api.GetBufferAndViewport();
+
+    // If we're not in the main buffer (either because an app has enabled the
+    // alternate buffer mode, or switched the conhost screen buffer), then VT
+    // paging doesn't apply, so we disregard the requested page number and just
+    // use the visible buffer (with a fixed page number of 1).
+    if (!isMainBuffer)
+    {
+        return { visibleBuffer, visibleViewport, 1 };
+    }
+
+    // If the requested page number happens to be the visible page, then we
+    // can also just use the visible buffer as is.
+    if (requestedPageNumber == _visiblePageNumber)
+    {
+        return { visibleBuffer, visibleViewport, _visiblePageNumber };
+    }
+
+    // Otherwise we're working with a background buffer, so we need to
+    // retrieve that from the buffer array, and resize it to match the
+    // active page size.
+    const auto pageSize = visibleViewport.size();
+    auto& pageBuffer = _getBuffer(requestedPageNumber, pageSize);
+    return { pageBuffer, til::rect{ pageSize }, requestedPageNumber };
+}
+
+Page PageManager::ActivePage() const
+{
+    return Get(_activePageNumber);
+}
+
+Page PageManager::VisiblePage() const
+{
+    return Get(_visiblePageNumber);
+}
+
+void PageManager::MoveTo(const til::CoordType pageNumber, const bool makeVisible)
+{
+    auto [visibleBuffer, visibleViewport, isMainBuffer] = _api.GetBufferAndViewport();
+    if (!isMainBuffer)
+    {
+        return;
+    }
+
+    const auto pageSize = visibleViewport.size();
+    const auto visibleTop = visibleViewport.top;
+    const auto wasVisible = _activePageNumber == _visiblePageNumber;
+    const auto newPageNumber = std::min(std::max(pageNumber, 1), MAX_PAGES);
+    auto redrawRequired = false;
+
+    // If we're changing the visible page, what we do is swap out the current
+    // visible page into its backing buffer, and swap in the new page from the
+    // backing buffer to the main buffer. That way the rest of the system only
+    // ever has to deal with the main buffer.
+    if (makeVisible && _visiblePageNumber != newPageNumber)
+    {
+        const auto& newBuffer = _getBuffer(newPageNumber, pageSize);
+        auto& saveBuffer = _getBuffer(_visiblePageNumber, pageSize);
+        for (auto i = 0; i < pageSize.height; i++)
+        {
+            saveBuffer.GetMutableRowByOffset(i).CopyFrom(visibleBuffer.GetRowByOffset(visibleTop + i));
+        }
+        for (auto i = 0; i < pageSize.height; i++)
+        {
+            visibleBuffer.GetMutableRowByOffset(visibleTop + i).CopyFrom(newBuffer.GetRowByOffset(i));
+        }
+        _visiblePageNumber = newPageNumber;
+        redrawRequired = true;
+    }

```

Comment:
More importantly, I'm a little worried about this change. Why don't we swap the visible buffer in the `_api` here? Is it because of some fundamental reason or did you do it due to our current, rather "intertwined" architecture?

If it's the latter I'm afraid this PR is making things worse. If I'm reading it correctly, the original DECs did not seem to have made a distinction between the "main visible" buffer and other pages. They were all pages, but one of them was simply the visible. Is my understanding correct? If so, do you think it's feasible to change the architecture in this PR? I'd completely understand if that was too difficult or time consuming.

Reply:
> Why don't we swap the visible buffer in the `_api` here?

Well the first problem is that anything in the `_api` needs to be written twice, because the buffer management in Windows Terminal is entirely separate from conhost. In an ideal world we'd have a unified buffer API that covered all the needs of the legacy console API, the XTerm alt buffers, and the VT paging, and which was shared between conhost and WT, but addressing that problem is way more work than I wanted to take on now.

And besides that, I'm not sure there'd be any benefit to it anyway. When swapping the visible page, it wouldn't just be a matter of pointing the renderer at the appropriate buffer - it's kind of mix. The scrollback (the `S` in that diagram above) is shared between all pages. It's only the `P+V` part that changes when you switch the visible page. 

> the original DECs did not seem to have made a distinction between the "main visible" buffer and other pages.

That is correct, but they didn't support scrollback and paging at the same time, so the `S` section in the diagram above would not have existed when paging was enabled. But modern terminals do tend to support both, so I thought it would be best if we could handle that too. And if you want to retain scrollback (which is stored in the main buffer) when switching the visible page, there's got to be some form of buffer merging.

Reply:
Yeah, that makes sense to me, at least from a VT point of view. And if that split also enabled us to have the viewport/page storage being a different width to the scrollback buffer, that might be a convenient way to support VT windowing (not an essential feature, but it's something which might be nice to have one day).

The things I think would be most problematic would be the legacy console APIs, and buffer resizing. But if you want to support infinite scrollback, those areas are likely to be complicated regardless of the architecture.